dht-cache/src/cache.rs

```diff
@@ -0,0 +1,462 @@
+                    }
+
+                    None
+                }
+                Event::Discovered(who) => Some(DomoEvent::NewPeers(
+                    who.into_iter().map(|w| w.to_string()).collect(),
+                )),
+                Event::VolatileData(data) => {
+                    // TODO we swallow errors quietly here
+                    serde_json::from_str(&data)
+                        .ok()
+                        .map(DomoEvent::VolatileData)
+                }
+                Event::PersistentData(data) => {
+                    if let Ok(mut elem) = serde_json::from_str::<DomoCacheElement>(&data) {
+                        if elem.republication_timestamp != 0 {
+                            log::debug!("Retransmission");
+                        }
+                        // TODO: do something with this value instead
+                        elem.republication_timestamp = 0;
+                        local_write
+                            .try_put(&elem)
+                            .await
+                            .ok()
+                            .map(|_| DomoEvent::PersistentData(elem))
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+    });
+
+    (cache, events)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::dht::test::*;
+    use std::{collections::HashSet, pin::pin};
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+    async fn builder() {
+        let cfg = crate::Config {
+            shared_key: "d061545647652562b4648f52e8373b3a417fc0df56c332154460da1801b341e9"
+                .to_owned(),
+            ..Default::default()
+        };
+
+        let (_cache, _events) = Builder::from_config(cfg).make_channel().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn syncronization() {
+        let [mut a, mut b, mut c] = make_peers(2).await;
+        let mut d = make_peer(2).await;
+
+        connect_peer(&mut a, &mut d).await;
+        connect_peer(&mut b, &mut d).await;
+        connect_peer(&mut c, &mut d).await;
+
+        let a_local_cache = LocalCache::new();
+        let b_local_cache = LocalCache::new();
+        let c_local_cache = LocalCache::new();
+        let d_local_cache = LocalCache::new();
+
+        let mut expected: HashSet<_> = (0..10)
+            .into_iter()
+            .map(|uuid| format!("uuid-{uuid}"))
+            .collect();
+
+        let (a_c, a_ev) = cache_channel(a_local_cache, a, 100);
+        let (b_c, b_ev) = cache_channel(b_local_cache, b, 100);
+        let (c_c, c_ev) = cache_channel(c_local_cache, c, 100);
+
+        let mut expected_peers = HashSet::new();
+        expected_peers.insert(a_c.peer_id.clone());
+        expected_peers.insert(b_c.peer_id.clone());
+        expected_peers.insert(c_c.peer_id.clone());
+
+        tokio::task::spawn(async move {
```

Comment:
I think we should avoid spawning tasks in tests, and "just" joining them. Spawning tasks in the runtime could make edge cases work by chance that otherwise, if we are lucky enough, could more easily deadlock in single thread and be detected by tests.

Reply:
You are absolutely right.